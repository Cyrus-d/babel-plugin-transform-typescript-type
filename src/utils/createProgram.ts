import * as ts from 'typescript';
import { getCompilerOptions, sourceFileCacheInstance } from './SourceFileCache';

export const createProgram = (filePath: string) => {
  const tsconfig = getCompilerOptions();
  const program = ts.createProgram([filePath], tsconfig, {
    fileExists(fileName): boolean {
      const file = sourceFileCacheInstance.createOrUpdateSourceFile(fileName);
      if (!file) {
        return false;
      }

      return true;
    },
    getCanonicalFileName(_fileName: string): string {
      return _fileName;
    },
    getCurrentDirectory(): string {
      return '';
    },
    getDefaultLibFileName(): string {
      return ts.getDefaultLibFileName(tsconfig);
    },

    getNewLine(): string {
      return '\r\n';
    },
    getSourceFile(fileName): ts.SourceFile | undefined {
      return sourceFileCacheInstance.createOrUpdateSourceFile(fileName);
    },
    readFile(fileName: string): string {
      const src = sourceFileCacheInstance.createOrUpdateSourceFile(fileName);
      if (!src) return '';

      return src.text;
    },
    useCaseSensitiveFileNames(): boolean {
      return false;
    },
    writeFile: (_fileName, _content) => null,
  });

  return program;
};
