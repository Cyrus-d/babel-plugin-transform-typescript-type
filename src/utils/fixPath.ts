export const fixPath = (fileName: string) => fileName.split('/').join('\\');
