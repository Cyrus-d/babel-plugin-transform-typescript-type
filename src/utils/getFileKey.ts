export const getFileKey = (fileName: string) =>
  fileName
    .split('/')
    .join('_')
    .split('\\')
    .join('_');
