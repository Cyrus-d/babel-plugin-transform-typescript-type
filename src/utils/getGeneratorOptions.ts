import { types as t } from '@babel/core';
import { GeneratorOptions } from '../typings';

interface Option {
  [key: string]: any;
}

const setArrayOption = (prop: t.ObjectProperty, opts: Option) => {
  if (!t.isArrayExpression(prop.value)) return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.elements.reduce((arr, elm) => {
      if (t.isStringLiteral(elm)) arr.push(elm.value);

      return arr;
    }, [] as string[]);
  }
};

const setOtherOptions = (prop: t.ObjectProperty, opts: Option) => {
  if (
    !t.isNumericLiteral(prop.value) &&
    !t.isStringLiteral(prop.value) &&
    !t.isBooleanLiteral(prop.value)
  )
    return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.value;
  }
};

export const getGeneratorOptions = <T extends t.CallExpression>(node: T): GeneratorOptions<T> => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (!node.arguments) return {} as GeneratorOptions<T>;

  return node.arguments.reduce((opts, arg) => {
    if (t.isObjectExpression(arg)) {
      arg.properties.forEach(prop => {
        if (t.isObjectProperty(prop)) {
          setArrayOption(prop, opts);
          setOtherOptions(prop, opts);
        }
      });
    }

    return opts;
  }, {} as GeneratorOptions<{}>);
};
