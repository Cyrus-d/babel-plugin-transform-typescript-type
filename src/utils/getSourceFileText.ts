import * as fs from 'fs';

export const getSourceFileText = (fileName: string) => {
  if (!fs.existsSync(fileName)) return null;

  return fs.readFileSync(fileName, 'utf8');
};
