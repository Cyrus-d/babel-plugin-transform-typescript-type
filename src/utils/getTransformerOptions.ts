import { types as t } from '@babel/core';
import { TransformerOptions } from '../typings';

interface Option {
  [key: string]: any;
}

const setArrayOption = (prop: t.ObjectProperty, opts: Option) => {
  if (!t.isArrayExpression(prop.value)) return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.elements.reduce((arr, elm) => {
      if (t.isStringLiteral(elm)) arr.push(elm.value);

      return arr;
    }, [] as string[]);
  }
};

const setOtherOptions = (prop: t.ObjectProperty, opts: Option) => {
  if (
    !t.isNumericLiteral(prop.value) &&
    !t.isStringLiteral(prop.value) &&
    !t.isBooleanLiteral(prop.value)
  )
    return;

  if (t.isIdentifier(prop.key)) {
    opts[prop.key.name] = prop.value.value;
  }
};

export const getTransformerOptions = <T extends t.CallExpression>(
  node: T,
): TransformerOptions<T> => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  if (!node.arguments) return {} as TransformerOptions<T>;

  return node.arguments.reduce((opts, arg) => {
    if (t.isObjectExpression(arg)) {
      arg.properties.forEach(prop => {
        if (t.isObjectProperty(prop)) {
          setArrayOption(prop, opts);
          setOtherOptions(prop, opts);
        }
      });
    }

    return opts;
  }, {} as TransformerOptions<{}>);
};
