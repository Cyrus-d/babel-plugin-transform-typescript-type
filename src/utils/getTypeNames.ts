import { types as t } from '@babel/core';

const getTSTypeReferenceTypeName = (node: t.TSTypeReference) => {
  if (t.isIdentifier(node.typeName)) {
    return node.typeName.name;
  }

  return null;
};

const getIntersectionTypeNames = (node: t.TSIntersectionType) => {
  const typeNames = node.types.reduce((arr, type) => {
    if (t.isTSTypeReference(type)) {
      const typeName = getTSTypeReferenceTypeName(type);
      if (typeName) arr.push(typeName);
    }

    return arr;
  }, [] as string[]);

  return typeNames;
};

export const getTsTypeName = (node: t.TSType) => {
  let typeNames: string[] = [];
  if (t.isTSTypeReference(node)) {
    const typeName = getTSTypeReferenceTypeName(node);
    if (typeName) typeNames.push(typeName);
  } else if (t.isTSIntersectionType(node)) {
    typeNames = getIntersectionTypeNames(node);
  }

  return typeNames;
};

const getTypeParametersTypeNames = (typeParameters: t.TSTypeParameterInstantiation) => {
  return typeParameters.params.reduce((arr, param) => {
    return getTsTypeName(param);
  }, [] as string[]);
};

export const getNodeTypesNames = (node: t.Node) => {
  if (t.isCallExpression(node)) {
    return getTypeParametersTypeNames((node as any).typeParameters);
  }

  return [];
};
