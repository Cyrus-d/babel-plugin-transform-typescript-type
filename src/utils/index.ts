export * from './getGeneratorOptions';
export * from './getTypeNames';
export * from './mergeSchema';
