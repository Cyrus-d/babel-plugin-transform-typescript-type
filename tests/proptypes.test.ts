import { transform } from './utils';
import path from 'path';

describe('prop types', () => {
  it('should transforms when is not in production', () => {
    expect(
      transform(
        path.join(__dirname, './fixtures/manual-converter/prop-types/component-prop-schema.ts'),
        {},
        {
          isProduction: false,
        },
      ),
    ).toMatchSnapshot();
  });

  it('should not transforms when is in Production ', () => {
    expect(
      transform(
        path.join(__dirname, './fixtures/manual-converter/prop-types/component-prop-schema.ts'),
        {},
        {
          isProduction: true,
        },
      ),
    ).toMatchSnapshot();
  });

  it('should not transforms when generateReactPropTypesManually=true ', () => {
    expect(
      transform(
        path.join(__dirname, './fixtures/manual-converter/prop-types/component-prop-schema.ts'),
        {},
        {
          generateReactPropTypesManually: true,
        },
      ),
    ).toMatchSnapshot();
  });

  it('should transforms when generateReactPropTypesInProduction=true', () => {
    expect(
      transform(
        path.join(__dirname, './fixtures/manual-converter/prop-types/component-prop-schema.ts'),
        {},
        {
          generateReactPropTypesInProduction: true,
          isProduction: true,
        },
      ),
    ).toMatchSnapshot();
  });

  it('should generate prop-types manually', () => {
    expect(
      transform(
        path.join(
          __dirname,
          './fixtures/manual-converter/prop-types/component-prop-schema-manual.ts',
        ),
        {},
        {
          generateReactPropTypesManually: true,
        },
      ),
    ).toMatchSnapshot();
  });
});
