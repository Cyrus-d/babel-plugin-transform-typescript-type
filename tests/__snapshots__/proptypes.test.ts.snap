// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prop types should generate prop-types manually 1`] = `
"import _pt from 'prop-types';
import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
import { generateComponentPropTypes } from '../../../../src/generatorFunctions';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.propTypes = {
  prop_a: _pt.string.isRequired,
  prop_b: _pt.bool
};"
`;

exports[`prop types should not transforms when generateReactPropTypesManually=true  1`] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};"
`;

exports[`prop types should not transforms when is in Production  1`] = `
"import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};"
`;

exports[`prop types should transforms when generateReactPropTypesInProduction=true 1`] = `
"import _pt from 'prop-types';
import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.propTypes = {
  prop_a: _pt.string.isRequired,
  prop_b: _pt.bool
};"
`;

exports[`prop types should transforms when is not in production 1`] = `
"import _pt from 'prop-types';
import React, { Component, SFC, memo, forwardRef, Ref } from 'react';
export interface ComponentProps {
  prop_a: string;
  prop_b?: boolean;
}

const FunctionComponent = (props: ComponentProps) => {
  return null;
};

FunctionComponent.propTypes = {
  prop_a: _pt.string.isRequired,
  prop_b: _pt.bool
};"
`;
